%% ISB705: Advanced Systems Biology II
% *Maria Pires Pacheco, Patricia Martins Conde, Tamara Bintener, and Thomas 
% Sauter*

clear all % this clears all your workspace variables. It helps you to have a fresh start in your script

path = pwd; % save working path

addpath(genpath(pwd)) %add the current folder and all subfolders to the Matlab path

delete clone*.log %delete some files generated by cplex

cd(path)

prompt = ('Hi! What is your name?\n');
userISB705
%% 
% this part is not mandatory to run, leave it commented

% addpath(genpath('C:\Program Files\IBM\ILOG\CPLEX_Studio128\cplex\Matlab')) % add cplex to the path
% addpath(genpath('C:\cobratoolbox-master')) % add the cobra tool
% rmpath(genpath('C:\cobratoolbox-master\src\dataIntegration\transcriptomics\FASTCORE')) %remove Cobra fastcore
%%
load model_practicals.mat % load used to load previously saved .mat files
disp('Scroll down to see all the fields')
model % displays information about the model variable
%% *Question 1:* 
% How many fields are in the model variable? You can answer the question running 
% the Answer section. 

% ---- enter your code here, then run the Answer section to enter the answer ----

numel(fieldnames(model))
%% 
% *Answer*

%Ques(1)
%% *Question 2:*
% What is the number of genes in the model ? 

% ---- enter your code here, then run the Answer section to enter the answer ----

numel(model.genes)
%% 
% *Answer*

%Ques(2)
%% 
% *Additional notes*
% 
% As you might have noticed, there are many fields in the model variable that 
% have the same number of rows or columns and are linked to each other. Usually, 
% in a model, there are many fields that additionally describe the reactions, 
% metabolites, and genes and these field have the same dimensions. For example, 
% the _rxnNames_ field gives you the complete name of the reactions stored in 
% _rxns_, the _subSystems_ field contains the pathways to which each reaction 
% belong. For the metabolites, you have several field that give you their formulas 
% and external identifiers. Note that often in these models, not all the fields 
% are complete even though they should be.
% 
% The first entry in the _subSystems_ field correspond to the first reaction 
% in the field rxns:

model.rxns(1) % displays the first reaction
model.rxnNames(1) % gives you the name of the first reaction
model.subSystems(1) % only shows you the cell, not its contents
%% *Question 3*
% Obviously, there is more than 1 reaction per pathway and, therefore, the pathways 
% are repeated several times. 
% 
% How many unique pathways are in the model?

% ---- enter your code here, then run the Answer section to enter the answer ----
numel(unique(model.subSystems))

%% 
% *Answer*

%Ques(3)
%% 
% *Pathways*
% 
% Now that we know how many unique pathways there are, you can also determine 
% the number of reactions in each pathways by typing:

[unique_sub, ~, ub] = unique(model.subSystems);
sub_counts = histc(ub, 1:length(unique_sub));
T = table(unique_sub, sub_counts)% scroll down to see the number of reactions every pathway, alternatively you can also open the T variable
%% 
% We can also plot the number of reaction per pathway in a barplot:

[~,I] = sort(T.sub_counts,'descend'); %sort by descreasing pathways 
figure
bar(T.sub_counts(I))
xlim([1 numel(T.unique_sub)])
% xticklabels(T.unique_sub(I)) %enable this to see the names, the figure will be very small
xticks(1:numel(T.unique_sub))
xtickangle(60)
title('Number of reactions per pathway')
%% *Question 4*
% How many reactions belong to the  "Transport, extracellular" pathway?

% ---- enter your code here, then run the Answer section to enter the answer ----

sum(ismember(model.subSystems, 'Transport, extracellular'))

%% 
% *Answer*

%Ques(4)
%% Question 5
% The following code  returns a vector with only ones and zeros of the dimension 
% 10601 by 1 (10601 rows and 1 column). There is a 1 if a reaction is present 
% in the extracellular transport pathway.

ismember(model.subSystems, 'Transport, extracellular')
%% 
% To what correspond correpond the following command?

find(ismember(model.subSystems, 'Transport, extracellular'))
%% 
% *Answer*

%Ques(5)
%% Question 6
% Let's have a look at the metabolites, what is the 59th metabolite called?

% ---- enter your code here, then run the Answer section to enter the answer ----
model.mets(59)

%% 
% *Answer*

%Ques(6)
%%
model.metcomp = regexprep(model.mets, '(.*?)\[',''); % removes everything before the bracket, including the bracket
model.metcomp = regexprep(model.metcomp, ']',''); %remove the closing bracket
%% *Question 7*
% How many reactons take place in the cytoplasm?

% ---- enter your code here, then run the Answer section to enter the answer ----

sum(ismember(model.metcomp,'c'))
%% 
% *Answer*

%Ques(7)
%%
model.grRules(1) % is not associated to any genes
model.grRules(96) %is associated to two genes using a Boolean OR
model.grRules(279) %is associated to two genes using a Boolean AND
%% Question 8
% In a GPR expression with a BOOLEAN OR, if one of the genes is active and the 
% other is inactive, is the reaction active ? (Yes/No)
% 
% *Answer*

%Ques(8)
%% Question 9
% In a GPR expression with a BOOLEAN AND, if one of the genes is active and 
% the other is inactive, is the reaction active ? (Yes/No)
% 
% *Answer*

%Ques(9)
%% Question 10
% What about this expression? If Gene 4 is expressed and the others are not, 
% is the reaction active?
% 
% Gene 1 OR (Gene 2 AND Gene 3) OR Gene 4
% 
% *Answer*

%Ques(10)
%% Chapter 2: Context-specific models
% Our model has the lower bound _lb_ field, so let us create the _rev_ field 
% for FASTCC:

model.rev = model.lb < 0 % the rev field will be a logical array
A = fastcc_4_rfastcormics(model, 1e-4, 0); % running FASTCC. 1e-4 needs to be used to avoid errors with small numbers.
%% Question 11
% How do you call a model in which all the reactions are connected directly 
% or indirectly to the environment?

%Ques(11)
%% Question 12
% There is only one reaction in the model that is blocked. What is the name 
% (model.rxns) of this reaction? Be careful with the spelling.

% ---- enter your code here, then run the Answer section to enter the answer ----

setdiff(model.rxns,model.rxns(A))

%% 
% *Answer*

%Ques(12)
%% 
% *Creation of a consistent model*
% 
% We have previously obtained the vector _A_ containing the indices of the consistent 
% reactions in the model. We will use this vector to build a model without any 
% blocked reaction(s) called consistent_model. Therefore, we are going to remove 
% these reactions using the _removeRxns_ function from the model.

consistent_model = removeRxns(model, model.rxns(setdiff(1:numel(model.rxns),A))) % create a consistent model based on the vector A
%% 
% To check if the newly created model is consistent, we can run _fastcc_ again, 
% it will tell us that the input model is consistent:

fastcc_4_rfastcormics(consistent_model, 1e-4, 0);
%% 
% *Exchange reactions and medium constraints*
% 
% Exchange reactions are reactions that connect the system to the extracellular 
% environment noted [e]. Not every exchange reaction can carry a flux in every 
% condition. For example, if we take cells in a dish that are growing on a medium 
% X, the cells are only able to import metabolites that are also found in the 
% medium. To correctly model these cells, the model has to be constrained to only 
% take up these available metabolites. 
% 
% The first step in constraining the model with the medium is to find its exchange 
% reactions by using the function _findExcRxns_. Again, the output is a logical 
% array that corresponds to 1 and 0 for each reaction in the model.

[EX] = findExcRxns(consistent_model)
%% Question 13
% How many Exchange are in the model?.

% ---- enter your code here, then run the Answer section to enter the answer ----

consistent_model.rxns(EX)
sum(EX)

%% 
% *Answer*

%Ques(13)
%% 
% If you want to know what metabolites are used in the exchange reactions you 
% can print the reactions equation by using the _printRxnFormula_ function:

printRxnFormula(consistent_model, consistent_model.rxns(find(EX))); % shows ALL exchange reactions
%% 
% As you scroll through the equations, you see that there is only a metabolite 
% on one side of the equation and nothing on the other. This means that the metabolite 
% is secreted into the extracellular space (X -> ) or going out of the model. 
% Other metabolites can either be secreted or taken up from the extracellular 
% space (X <=> ).
% 
% Demand reactions are denoted by DM_ and sink reactions are denoted by sink_. 
% These reactions allow for a limitless production or consumption of a metabolite.
%% *Question 14*
% To constrain a model, for every metabolite that is not present in the medium, 
% the bounds of the exchange reactions that take up this metabolite has to be 
% set to 0.
% 
% First, import the provided example medium with its composition:

load medium_example.mat
%% 
% How many metabolites are present in the example medium?

% ---- enter your code here, then run the Answer section to enter the answer ----

length(medium_example)
%% 
% *Answer* 

%Ques(14)
%% *Question 15*
% Secondly, let us find the exchange reactions that are import these metabolites 
% from the cell into the model. These exchange reactions will be able to carry 
% a flux. We can use the _findRxnsFromMets_ function to find the reactions associated 
% to a list of metabolites:

[Ex_open] = findRxnsFromMets(consistent_model, medium_example)
%% 
% Find reactions that should be unable to carry a flux and save them in a variable 
% called  "_Ex_to_close"._

% ---- enter your code here, then run the Answer section to enter the answer ----

Ex_to_close = setdiff(consistent_model.rxns(EX), Ex_open)

%% 
% *Answer*

%Ques(15, Ex_to_close)
%% Question 16
% Do you have to constrain the upper or the lower bounds in the model? Type 
% _ub_ for upper bounds and _lb_ for lower bounds.

%Ques(16)
%% 
% Now adjust the following code to change these bounds to zero. Replace XX by 
% either ub or lb. This will change the bound to 0.

medium_constrained_model = consistent_model; %let us save the consistent model in case you make a mistake
medium_constrained_model.lb(ismember(consistent_model.rxns,Ex_to_close)) = 0
%% 
% After changing the bounds of these reactions, they will no longer be able 
% to carry a flux and the model might become inconsistent again. We should run 
% FASTCC again on the medium constrained model:

A = fastcc_4_rfastcormics(medium_constrained_model, 1e-4, 0); %takes approx. 15 minutes
disp('DONE!')
%% Question 17
% After getting the indices of the consistent reactions from the medium constrained 
% model, build the new consistent and medium constrained model by removing the 
% inconsitent reactions. Name the model  medium_constrained_consistent_model

% ---- enter your code here, then run the Answer section to enter the answer ----

medium_constrained_consistent_model = removeRxns(medium_constrained_model, medium_constrained_model.rxns(setdiff(1:numel(medium_constrained_model.rxns),A)))
%% 
% What is the size of the medium_constrained_consistent_model (number of reactions)?

% ---- enter your code here, then run the Answer section to enter the answer ----

length(medium_constrained_consistent_model.rxns)
%% 
% *Answer* 

%Ques(17)
%% *2) Reconstruction of context-specific models using gene expression data*
%% 
% *Example*
% 
% In the following section we will get some hands-on on how these algorithms 
% work. Therefore, we are going to use the provided _data.mat_ that contains RNA-seq 
% data from 25 different breast cancer samples from the TCGA dataset. The data.mat 
% file contains the variables _fpkm_, _colnames_, and _rownames_. Have a look 
% at each variable to know what it contains.
% 
% In a first step, we will load and visualize the fpkm data. 

load data.mat % contains fpkm, colnames, and rownames
fpkm = table2array(fpkm); %transform table to array
%% 
% The FPKM data is one way on how to process raw RNA-seq data. FPKM stands for 
% Fragments Per Kilobase of transcript per Million mapped reads. In layterms one 
% can say that it measures the activity of a gene. The higher the values, the 
% more likely it is that the given gene is expressed in the sample. You can find 
% more information on RNA-seq data here: <https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/ 
% https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/> 
% 
% Because the distribution of the data is highly skewed, we will first log transform 
% the data and then remove all the -Inf values.

signal = log2(fpkm(:,1)); % log2-transform the first sample of the fpkm data
signal(isinf(signal)) = -100000; %find the -Inf values and change them to -100000
signal = signal(signal>-10000); % remove samples with low expression
%% 
% Let us plot the distribution of the first sample

figure
hist((signal),100);
ylabel('signal'); xlabel('log2(FPKM)')
%% 
% Now that we have the distribution of the log transformed fpkm data, we need 
% to determine which genes are expressed and which genes are not expressed. We 
% are therefore going to use the proposed method from rFASTCORMICS that will be 
% called by using fittingcurve.

fittingcurve %operating on "signal"
%% 
% *Data discretization*

discretized = discretize_FPKM(fpkm, colnames);
%% Question 18
% Let's say a reaction is under the control of 4 genes:
%% 
% * gene 1 has an expression of 100 fpkm
% * gene 2 has an expression of 0.5 fpkm
% * gene 3 has an expression of 200 fpkm
% * gene 4 has an expression of 0.1 fpkm
%% 
% What will be the expression associated to the reaction
% 
% |(Gene 1 OR Gene) 2 AND (Gene 3 OR Gene 4)|

%Ques(18)
%% Question 19
% The same can be done with discretized data: 
%% 
% * gene 1 has a score of 1 (expressed)
% * gene 2 has a score of 0 (unknown expression)
% * gene 3 has a score of 1 (expressed)
% * gene 4 has a score of -1 (not expressed)
%% 
% |Gene 1 and  Gene 2 and Gene 3 or Gene 4| 
% 
% What will be the score associated to the reaction?

%Ques(19)
%% 3) Model reconstruction using rFASTCORMICS
%% 
% Create your breast cancer model based on the first column of the data and 
% the consistent model. We have provided the dico and some of the optional settings 
% for you. 

load optional_settings
load dico

already_mapped_tag = 0;
consensus_proportion = 0.9;
epsilon = 1e-4;
%% Question 20
% Use the _fastcormics_RNAseq_ function to create a generic context-specific 
% model for breast cancer using all the discretized samples. 
% 
% What is the size of the context-specific model (number of reactions)?

% ---- enter your code here, then run the Answer section to enter the answer ----
feature astheightlimit 2000

[~, A_final] = fastcormics_RNAseq(consistent_model, discretized, rownames, dico, already_mapped_tag, consensus_proportion, epsilon, optional_settings)
numel(A_final)
%% 
% *Answer*

%Ques(20)
%% 
% *Create a model from A_final*
% 
% Each tissue has a different expression pattern which is taken into account 
% by FASTCORMICS. In fact, compared to other model building algorithms, the FASTCORE 
% family has a high accuracy, specificity and is very robust. You can read more 
% on the performance of different algorithms here: <https://www.frontiersin.org/articles/10.3389/fphys.2015.00410/full 
% https://www.frontiersin.org/articles/10.3389/fphys.2015.00410/full> 
% 
% To create a model from the A_final array, we will create a logical matrix 
% with the samples as column and the reactions as row in which 1 depicts if a 
% reaction should take place in the context-specific model. We will call this 
% matrix: models_keep
% 
% First, we create an array of all zeros

models_keep = zeros(numel(consistent_model.rxns), 1) %all 0 array, with same size as reactions in the model
%% 
% using the indices in A_final, we will set to 1 those reactions that take place

models_keep(A_final,1) = 1;
%% 
% based on this array, we can create the models, by removing the unnecessary 
% reactions from the consistent model

BRCA_generic_model = removeRxns(consistent_model,consistent_model.rxns(setdiff(1:numel(consistent_model.rxns),find(models_keep(:,1)))))
%% 
% Note: the model is also a direct output from FASTCORMICS, however, if you 
% create many models it is better to suppress this output as it will only take 
% up memory and slow down your computer. Saving the model reaction in a matrix 
% is sufficient.
%% Model optimization
%% Question 21
% Often, an objective function is already set in the model. By default, it is 
% the biomass production, however it is always good to confirm that you are using 
% the correct objective function. You can either find it  by having a look the 
% model.c field where 1 marks the objective function reactions or by using the 
% _checkObjective_ function:

checkObjective(consistent_model)
find(model.c) % index of the objective function
%% 
% In this case, the objective of the model is the biomass maintenance reaction 
% with the index 5626 (Biomass maintenance reaction without replication precursors) 
% but we would like to change it to the complete biomass reaction. 
% 
% Find the index of the biomass reaction called "biomass_reaction"

% ---- enter your code here, then run the Answer section to enter the answer ----

idx = find(ismember(consistent_model.rxns,'biomass_reaction'))
%% 
% *Answer*

%Ques(21)
%% 
% To change the objective function in the model you can either change the model.c 
% field or use the changeObjective function. The output will be a struct/a new 
% model, so no not forget to save it.

consistent_model = changeObjective(consistent_model,consistent_model.rxns(idx));
%% 
% Now check if the objective function was changed successfully:

checkObjective(consistent_model)
%% Question 22
% In this example, we will use the biomass production as the objective function. 
% While optimizing for the biomass, we assume that the whole metabolism of a cell 
% is designed to produce biomass to enable fast proliferation.
% 
% For which of the following examples is this assumption reasonable:
%% 
% * a) neuron
% * b) cancer cell
% * c) E. coli 
% * d) none of the above 
% * e) all the above 
% * f ) a and b
% * g) b and c
% * h) a and c
%% 
% For which cells is the optimization of the biomass production reasonable? 
% (a, b, c, d, e, f, g,h)
% 
% *Answer*

%Ques(22)
%% 
% *Optimization*
% 
% We can now optimization the model for its biomass production to get the maximal 
% value through the biomass reaction. Note that we use the option 'max', because 
% the want to obtain the flux distribution that allows the highest production 
% of biomass. For other optimizations,  minimization might be a better option.

changeCobraSolver('ibm_cplex');
solution = optimizeCbModel(consistent_model,'max')
%% 
% The solution variable is again a structure with different fields. The field 
% we are interested in is the solution.f field, it gives you the maximal value 
% of the flux through the objective function. This value will be used in the single 
% gen deletion for the wild type.
%% Question 23
% What is this maximal growth value for the consistent model?

%Ques(23)
%% In _silico_ gene deletions
% Let's us perform singeGeneDeletion on the consistent model, while optimizing 
% for the biomass reaction to see what gene deletions affect the biomass.

[grRatio_consistent, grRateKO_consistent, grRateWT_consistent, ~, ~, ~, geneList_consistent] = singleGeneDeletion_MISB(consistent_model, 'FBA', [], 0, 1);
%% 
% Now we will perform the single gene deletion on the _medium-constrained consistent 
% model_ and the generic breast cancer models. Do not forget to check for the 
% objective function and change accordingly.
%% Question 24
% For the context-specific BRCA model:

BRCA_generic_model = changeObjective(BRCA_generic_model,consistent_model.rxns(idx));
checkObjective(BRCA_generic_model)
[grRatio_BRCA, grRateKO_BRCA, grRateWT_BRCA, ~, ~, ~, geneList_BRCA] = singleGeneDeletion_MISB(BRCA_generic_model, 'FBA', [], 0, 1);
%% 
% How many gene deletions reduce the biomass in the breast cancer model, i.e. 
% reduce the biomass to at least 90%.

% ---- enter your code here, then run the Answer section to enter the answer ----

sum(grRatio_BRCA < 0.9)
%% 
% *Answer*
% 
% How many gene deletions reduce the biomass in the breast cancer model, i.e. 
% reduce the biomass to at least 90%.

%Ques(24)
%% Question 25
% How many genes completely reduce the flux through the biomass reaction to 
% 0 in the breast cancer model?

% ---- enter your code here, then run the Answer section to enter the answer ----
sum(grRatio_BRCA == 0)

%% 
% *Answer*

%Ques(25)
%% 
% *Visualization*
% 
% The growth ratios can easily be plotted to visualize how many genes affect 
% the objective function. You will have to sort the growth ratios before plotting 
% them.

figure
% use plot to plot the growth ratios in ascending order

plot(sort([grRatio_consistent, grRatio_BRCA],'ascend'))

title('Sorted growth ratios for the models')
xlabel('number of genes')
ylabel('growth ratios')
xlim([0 200])
ylim([-0.1 1.1])
legend({'consistent model','breast cancer model'},'Location','best')
%% Essential genes

threshold = 0.5
essential_genes_consistent = grRatio_consistent <= threshold;
essential_genes_BRCA = grRatio_BRCA <= threshold;
%% Question 26
% For a threshold of 0.5, how many genes are essential in the consistent model?

% ---- enter your code here, then run the Answer section to enter the answer ----

sum(essential_genes_consistent)
%% 
% *Answer*

%Ques(26)
%% Question 27
% For a threshold of 0.5, how many genes are essential in the breast cancer 
% model?

% ---- enter your code here, then run the Answer section to enter the answer ----

sum(essential_genes_BRCA)
%% 
% *Answer*

%Ques(27)